### PR TITLE
wasm: configuration implementation for WAF extension

### PR DESCRIPTION
**Pull request by adeladeng**
Commit Message: Modified the example wasm modules to work as a WAF extension with configuration. The extension can detect SQL injection in http requests based on ModSecurity rules.
Risk Level: Medium
Testing: No testing has been done yet
Docs Changes: New README file at the root of the repository for building and running the module

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,93 +1,56 @@
...
+The source code for the WASM extension is in `examples/wasm`. Build the WASM module with:
+
+```bazel build //examples/wasm:envoy_filter_http_wasm_example.wasm```
+
+The WASM binary being built will be at
+`bazel-bin/examples/wasm/envoy_filter_http_wasm_example.wasm`. Make sure that the `filename` path in `examples/wasm/envoy.yaml` matches the path to the WASM binary. Then run the WASM module:
+
+``` sudo bazel-bin/source/exe/envoy-static -l trace --concurrency 1 -c `` `pwd`/examples/wasm/envoy.yaml`` ```
+
+In a separate terminal, curl at `localhost:80` to interact with the running proxy.
```
Comment (fpesce) [inline README.md]: I'd recommend to give the example of curl command that you run here.

File: README.md [hunk 2]
```diff
@@ -1,93 +1,56 @@
...
+## Deployment
+From the root of the repository, build static binary of envoy proxy:
+
+```bazel build -c opt //source/exe:envoy-static```
+
+Run tests for envoy to make sure the binary has been built successfully:
+
+```bazel test //test/common/common/...```
+
+The source code for the WASM extension is in `examples/wasm`. Build the WASM module with:
```
Comment (fpesce) [inline README.md:33]: You probably need to specify at this point that people running this command need to first have checked out libinjection repository in upper directory.

File: examples/wasm/envoy-config.yaml [hunk 3]
```diff
@@ -0,0 +1,91 @@
+static_resources:
+  listeners:
+  - name: main
+    address:
+      socket_address:
+        address: 0.0.0.0
+        port_value: 80
```
Comment (fpesce) [inline examples/wasm/envoy-config.yaml]: make this port anything above 1024 (actually in 80xx space, 8000 should be good), so you don't need to be root to run the envoy command.

File: README.md [hunk 4]
```diff
@@ -1,93 +1,56 @@
...
+
+```bazel build //examples/wasm:envoy_filter_http_wasm_example.wasm```
+
+The WASM binary being built will be at
+`bazel-bin/examples/wasm/envoy_filter_http_wasm_example.wasm`. Make sure that the `filename` path in `examples/wasm/envoy.yaml` matches the path to the WASM binary. Then run the WASM module:
+
+``` sudo bazel-bin/source/exe/envoy-static -l trace --concurrency 1 -c `` `pwd`/examples/wasm/envoy.yaml`` ```
```
Comment (fpesce) [inline README.md]: don't need the `sudo` if you follow my recommandation for `examples/wasm/envoy-config.yaml`

**Review by fpesce — CHANGES_REQUESTED** [review]
First round of comments (I haven't reviewed the code of the wasm module yet, doing that next)

File: examples/wasm/envoy_filter_http_wasm_example.cc [hunk 5]
```diff
@@ -1,14 +1,219 @@
...
+/*
+ * Keep track of config info
+ * Three main fields: param(query param), header, cookie
+ *
+ * If <field>_include is true, <field>s contains the names to be included in
+ * sql injection detection
+ * If <filed>_include is false, <field>s contains the names to be excluded
+ */
+struct Config {
+  std::string content_type { "application/x-www-form-urlencoded" };
```
Comment (fpesce) [inline examples/wasm/envoy_filter_http_wasm_example.cc]: Maybe you want to have a const string somewhere for "application/x-www-form-urlencoded" as you use this here and  on lines +270 and +347

File: examples/wasm/envoy_filter_http_wasm_example.cc [hunk 6]
```diff
@@ -1,14 +1,219 @@
...
+  return str;
+}
+
+/*
+ * Detect SQL injection on given input string
+ */
+bool detectSQLi(std::string input, std::string key, std::string part) {
+  struct libinjection_sqli_state state;
+  char* input_char_str = const_cast<char*>(input.c_str());
+  libinjection_sqli_init(&state, input_char_str, input.length(), FLAG_NONE);
```
Comment (fpesce) [inline examples/wasm/envoy_filter_http_wasm_example.cc:156]: Does that need to be released/free()'d somewhere?

**Review by fpesce — CHANGES_REQUESTED** [review]

File: examples/wasm/envoy_filter_http_wasm_example.cc (hunk 6 above)
Comment (adeladeng) [reply examples/wasm/envoy_filter_http_wasm_example.cc:156]: I don't think the struct needs to be freed, but I am not sure. I copied this declaration from the sample usage in libinjection repository. The sample code is in C, and in this syntax the struct will be allocated on stack and will be destroyed after the function exits, so there's no need to free after use. Does C++ work the same way?

Comment (adeladeng) [reply examples/wasm/envoy_filter_http_wasm_example.cc:156]: I will try declaring state with "new" and then delete it after use. Will that be more preferable in C++?

File: README.md (hunk 1 above)
Comment (adeladeng) [reply README.md]: Updated in Commit [here](https://github.com/dengyijia/envoy-wasm/pull/1/commits/288988de7a61647c396af164445775ef2511bf75)

File: examples/wasm/envoy_filter_http_wasm_example.cc (hunk 5 above)
Comment (adeladeng) [reply examples/wasm/envoy_filter_http_wasm_example.cc]: Updated in commit [here](https://github.com/dengyijia/envoy-wasm/pull/1/commits/340940bd7a1de29740d59ef6f8e1199b4b1fa1ba). I am not sure of the recommended way to add a global variable in C++. Should I declare it within some class?

File: examples/wasm/envoy-config.yaml (hunk 3 above)
Comment (adeladeng) [reply examples/wasm/envoy-config.yaml]: Updated in this [commit](https://github.com/dengyijia/envoy-wasm/pull/1/commits/6cfcd9373cb609b95ec42ce16c6716fcb671a7a1)

File: README.md (hunk 4 above)
Comment (adeladeng) [reply README.md]: Updated in this [commit](https://github.com/dengyijia/envoy-wasm/pull/1/commits/6cfcd9373cb609b95ec42ce16c6716fcb671a7a1)

File: README.md (hunk 2 above)
Comment (adeladeng) [reply README.md:33]: Updated in [this commit](https://github.com/dengyijia/envoy-wasm/pull/1/commits/3570d0ceb24647fc8dbf3d43afa2e9c8fee9df90)

File: examples/wasm/envoy_filter_http_wasm_example.cc [hunk 7]
```diff
@@ -32,19 +243,122 @@ bool ExampleRootContext::onStart(size_t) {
...
   }
+
+  // record body content type to context
+  content_type = getRequestHeader("content-type")->toString();
+
   return FilterHeadersStatus::Continue;
 }
 
+FilterDataStatus ExampleContext::onRequestBody(size_t body_buffer_length, bool end_of_stream) {
+  auto body = getBufferBytes(WasmBufferType::HttpRequestBody, 0, body_buffer_length);
```
Comment (yangminzhu) [inline examples/wasm/envoy_filter_http_wasm_example.cc:345]: don't you need to cache the body and execute the detectSQLiOnParams() on all the buffered body once?

File: examples/wasm/envoy_filter_http_wasm_example.cc [hunk 8]
```diff
@@ -32,19 +243,122 @@ bool ExampleRootContext::onStart(size_t) {
...
+  auto body_str = std::string(body->view());
+  LOG_ERROR(std::string("onRequestBody ") + body_str);
+
+  if (content_type.compare("application/x-www-form-urlencoded") != 0) {
+    return FilterDataStatus::Continue;
+  }
+
+  // detect SQL injection in query parameters
+  auto query_params = parseBody(body_str);
```
Comment (yangminzhu) [inline examples/wasm/envoy_filter_http_wasm_example.cc:354]: it seems strange to call parseBody but then store it in `query_params`.

File: examples/wasm/envoy_filter_http_wasm_example.cc [hunk 9]
```diff
@@ -22,6 +229,10 @@ class ExampleContext : public Context {
   void onDone() override;
   void onLog() override;
   void onDelete() override;
+
+private:
+  std::string content_type;
+  struct Config config;
```
Comment (yangminzhu) [inline examples/wasm/envoy_filter_http_wasm_example.cc]: `config_` for class members, same for others.

File: examples/wasm/envoy_filter_http_wasm_example.cc [hunk 10]
```diff
@@ -32,19 +243,122 @@ bool ExampleRootContext::onStart(size_t) {
   return true;
 }
 
-void ExampleContext::onCreate() { LOG_WARN(std::string("onCreate " + std::to_string(id()))); }
+bool ExampleRootContext::onConfigure(size_t config_size) {
```
Comment (yangminzhu) [inline examples/wasm/envoy_filter_http_wasm_example.cc:246]: will it have problems if this function is called multiple times and the same config is reused? or is this function only called once?

File: examples/wasm/envoy_filter_http_wasm_example.cc [hunk 11]
```diff
@@ -1,14 +1,221 @@
...
+  return false;
+}
+
+/**
+ * Detect SQL injection on given parameter pairs with configuration
+ * Input
+ *  - params: a map of param key value pairs
+ *  - include: a boolean
+ *      if true, given keys are the only keys to detect
+ *      if false, given keys are all but the given keys will be detected
```
Comment (yangminzhu) [inline examples/wasm/envoy_filter_http_wasm_example.cc]: the comment needs a bit update, not clear what does it mean: ` given keys are all but the given keys will be detected`?

File: examples/wasm/envoy_filter_http_wasm_example.cc (hunk 10 above)
Comment (adeladeng) [reply examples/wasm/envoy_filter_http_wasm_example.cc:246]: onConfigure is called only once when the module starts running. onCreate is called upon every new request.

File: examples/wasm/envoy_filter_http_wasm_example.cc (hunk 8 above)
Comment (adeladeng) [reply examples/wasm/envoy_filter_http_wasm_example.cc:354]: Should I rename the variable/function names to make it more intuitive? I thought the parameters passed in through the body are generally called "query parameters".

File: examples/wasm/envoy_filter_http_wasm_example.cc (hunk 9 above)
Comment (adeladeng) [reply examples/wasm/envoy_filter_http_wasm_example.cc]: Updated in [this commit](https://github.com/dengyijia/envoy-wasm/pull/1/commits/b44fa56d662316f60294b4745363e7ea67139943).

File: examples/wasm/envoy_filter_http_wasm_example.cc (hunk 7 above)
Comment (adeladeng) [reply examples/wasm/envoy_filter_http_wasm_example.cc:345]: Does "cache the body" mean store the body in some fields under ExampleContext? I assumed that the body will be passed in with valid url encoding, so I just need to decode and detect SQL injection on individual parameters in the body.

File: examples/wasm/envoy_filter_http_wasm_example.cc (hunk 11 above)
Comment (adeladeng) [reply examples/wasm/envoy_filter_http_wasm_example.cc]: That was a typo! Thanks for catching it. I was hoping to say 
```
if true, given keys are the only keys to detect
if false, all but the given keys will be detected
``` 
Do I need to further explain that?

File: examples/wasm/envoy_filter_http_wasm_example.cc (hunk 6 above)
Comment (fpesce) [reply examples/wasm/envoy_filter_http_wasm_example.cc:156]: Quickly looking at libinjection_sqli_init() source code, I don't seem to see a need for any freeing method here. So we;re good.

**Review by fpesce — APPROVED** [review]

**Pull request merged**